writer.go

```diff
@@ -0,0 +1,565 @@
+	defer func() {
+		if err != nil {
+			for _, index := range d.indexes {
+				if index.db != nil {
+					index.db.Close()
+				}
+			}
+		}
+	}()
+
+	// Setup primary table.
+	if err := d.setupIndex(d.primary); err != nil {
+		return err
+	}
+
+	// Setup index tables.
+	for _, index := range d.indexes {
+		if err := d.setupIndex(index); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (d *Writer) setupIndex(index *Index) error {
+	badgeOpts := writeOpts.WithDir(index.path).WithValueDir(index.path).WithLogger(d.logger)
+	if d.inmemory {
+		badgeOpts = badgeOpts.WithInMemory(true)
+	}
+
+	table, err := badger.Open(badgeOpts)
+	if err != nil {
+		return err
+	}
+
+	index.db = table
+	return nil
+}
+
+// poolMember is a goroutine that takes functions off the channel and executes them.
+func (d *Writer) poolMember(ch chan func()) {
+	for f := range ch {
+		f()
+	}
+}
+
+// Close closes out the Writer.
+func (d *Writer) Close() error {
+	d.wg.Wait()      // Wait for all writes to stop.
+	close(d.writeCh) // Kill all goroutines.
+
+	if len(d.primary.toWrite) != 0 {
+		err := d.primary.db.Update(
+			func(txn *badger.Txn) error {
+				for _, kv := range d.primary.toWrite {
+					if err := txn.Set(kv[0], kv[1]); err != nil {
+						return err
+					}
+				}
+				return nil
+			},
+		)
+		if err != nil {
+			return err
+		}
+	}
+
+	for { // This logic, which looks weird, comes from the Badger documentation.
+		err := d.primary.db.RunValueLogGC(0.7)
+		if err == nil {
+			continue
+		}
+		break
+	}
+
+	if err := d.primary.db.Flatten(runtime.NumCPU()); err != nil {
+		return err
+	}
+
+	if err := d.primary.db.Close(); err != nil {
+		return err
+	}
+
+	def := tableDef{Columns: make([]columnDef, 0, len(d.indexes)), Length: d.primary.counter.Load()}
+
+	for _, index := range d.indexes {
+		// If we still have values to write to our indexes, do that.
+		if len(index.toWrite) != 0 {
+			err := index.db.Update(
+				func(txn *badger.Txn) error {
+					for _, kv := range index.toWrite {
+						if err := txn.Set(kv[0], kv[1]); err != nil {
+							return err
+						}
+					}
+					return nil
+				},
+			)
+			if err != nil {
+				return err
+			}
+		}
+
+		def.Columns = append(def.Columns, columnDef{Name: index.Name, Path: index.path, AllowDuplicates: index.AllowDuplicates})
+
+		for { // This logic, which looks weird, comes from the Badger documentation.
+			err := index.db.RunValueLogGC(0.7)
+			if err == nil {
+				continue
+			}
+			break
+		}
+
+		if err := index.db.Flatten(runtime.NumCPU()); err != nil {
+			return err
+		}
+
+		if err := index.db.Close(); err != nil {
+			return err
+		}
+	}
+
+	return def.marshal(d.dirPath)
+}
+
+// Write data writes data to our database. indexes must be in the same order when you created
+// this DB and have the same number of indexes. You cannot reuse any "value" or "indexValues" passed until
+// all data has been written. This is because a single WriteData() does not cause data to be written.
```

Comment:
longer term, it would probably be beneficial to create a container to store an arbitrary set of indexes in an order, so you can create one, and use it to both create the DB and write data to it without having to remember the ordering

Reply:
I'm just not sure you can do this with a []byte type of data.  Since the value and indexValues are simply bytes, even the container type would end up doing exactly this.  I think you'd end up with a level of abstraction that looked identical in the end.

In reality, this is exactly how non-structured data stuff works in the standard lib, like the csv package.  You have to keep the order.  

Now, if I built something bringing structured data in on top of this, I could have that simply take something like a proto or struct definition in and then use protoreflect/reflect to then always keep the order and the user would never have to.  

However, this may just be my lack of vision, if you see a way to do this with plain bytes, I could incorporate it.

Reply:
I still don't see what the harm is in creating a `type IndexValues` (or a better name) that contains a single `[][]byte` field in it. you're right that it would not add any new functionality, but if this function had a signature that looked more like `(value []byte, indices IndexValues)`, which would force the caller to put their slice of `[]byte` into an order before they call this function. This really only matters if the create function (I'm presuming this is `WithIndexes`) takes the same or a related type, because the goal was to force callers to define the order once across the two different function calls. So, maybe a better way is to make this `[]byte` value derivable from `WithIndexes`'` `[]Index` value.

Reply:
I've made an update that I think addresses what your looking for.  Have a look